Improve BiomeDictionary's interaction with the new registries

diesieben07 opened the pull request:
Main point here is that `BiomeDictionary.registerBiomeType` no longer silently ignores unregistered biomes without any kind of indication.
At the same time update the class to decouple from vanilla's registries and use the cleaner forge interface.
This now makes the return type redundant, should I change as part of the "1.11.x binary change"?

There is a lot more that's wrong with the class:

- The formatting is all over the place and does not follow forge's standards.
- There are unused, seemingly random methods like `registerAllBiomes`, which does nothing but log a warning.

Should this be cleaned up?

diesieben07 committed:
Improve BiomeDictionary's interaction with the registry

diesieben07 committed:
Add two more places

mezz commented:
Please do clean this up, breaking changes are fine right now.

diesieben07 commented:
After digging more into this class, there is a lot more I found...

- Method names that remind me of MCP like `BiomeDictionary.getTypesForBiome` (instead of `getTypes`)
- Using an Enum and `EnumHelper` to add new instances instead of a proper design. `EnumHelper` should only be used where we really need to because of vanilla, right?
- Sillynesses like an `ArrayList<X>[]` with enum ordinals as the index (and then lord behold if someone adds a new enum entry, see above) instead of a `Map`.

How far do you want me to go?

williewillus commented:
go big or go home :D

mezz commented:
If you want to make it cleaner, go for it!
Just don't break the interface badly if you don't have to, renaming those methods is ok and removing the literally useless method is ok. 

Some stuff in there is pretty slow as well, like `isBiomeOfType`. Don't bother optimizing if you have to sacrifice readability or create a crazy data structure though.

diesieben07 committed:
- stop abusing Enums as a named-object structure, this improves and simplifies a lot of the data structures that are in place
- registerBiomeType no longer returns boolean (was always true)
- remove useless registerAllBiomes method
- Use proper collections instead of copying arrays everywhere
- Update formatting

diesieben07 commented:
Theres my take at it, let me know if I missed anything or if there is anything to improve about my changes.

diesieben07 committed:
missed one final

bs2609 commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -21,132 +21,133 @@
             {
-                if (t.name().equals(name))
-                    return t;
+                t = new Type(name, subTypes);
```
should there be a `byName.put(name, t)` here?

bs2609 commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -160,41 +161,30 @@ public BiomeInfo(Type[] types)
-                {
-                    getBiomeInfo(biome).typeList.add(type);
-                }
+                getBiomeInfo(biome).types.add(type);
```
foreach could be replaced with `getBiomeInfo(biome).types.addAll(subTags);`

diesieben07 replied:
Fixed.

diesieben07 committed:
Use addAll instead of manually adding every element

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -21,132 +21,133 @@
+         * NO Special characters. {Unenforced, just don't be a pain, if it becomes a issue I WILL
+         * make this RTE with no worry about backwards compatibility}
+         * <p>
          * Note: For performance sake, the return value of this function SHOULD be cached.
```
Does this performance warning still apply?

diesieben07 replied:
Probably not as much anymore, since it's just a hash lookup now. But if you are calling this a lot during worldgen it might be better to put it in a static final field, if only for cleaner code.

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -203,14 +190,9 @@ public static boolean registerBiomeType(Biome biome, Type ... types)
      * @param type the Type to look for
      * @return a list of biomes of the specified type, null if there are none
```
This `@return` javadoc needs to be updated

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -219,27 +201,27 @@ public static boolean registerBiomeType(Biome biome, Type ... types)
      * @param biome the biome to check
      * @return the list of types, null if there are none
```
This `@return` javadoc needs to be updated

diesieben07 committed:
Fix documentation on nullability

mezz commented:
Breaking things:
I'm thinking `registerAllBiomesAndGenerateEvents` could be pretty much inlined into `ForgeModContainer` where nobody else can call it.
It might be a good idea to log a short warning in `checkRegistration` to alert modders when their Biome has no Type set.

Style:
You use package-private a lot, which is weird. Please choose public, private, or protected like the rest of forge.

diesieben07 commented:
> You use package-private a lot, which is weird. Please choose public, private, or protected like the rest of forge.

I use package-private here for fields that need to be accessed in the outer class. E.g.  `Type::biomes` is accessed outside the `Type` class. Is that not the use for package-private?

mezz commented:
A class can access the private fields and methods of its inner classes. 
package-private is wider permission, I'm pretty sure anything in `net.minecraftforge.common` can access it.

williewillus commented:
Accessing the private members of inner classes actually just generates protected helper bridge functions iirc

mezz commented:
I went to read up on it a bit. In the bytecode, inner class get package-private accessors because the inner classes are put into their own files in the same package. I think this matches what williewillus said.
In the IDE it is still useful to have them as private to limit the scope.

diesieben07 commented:
I know, I know. I am pedantic about these things...

diesieben07 committed:
- make fields private instead of package
- move registerAllBiomesAndGenerateEvents to ForgeModContainer
- simplify logic in containsType
- rename checkRegistered->ensureRegistered to better reflect it's functionality

diesieben07 committed:
Cleanup method names

diesieben07 commented:
Addressed your remarks.

diesieben07 committed:
- unify wording regarding "tags" and "types"
- put initialization of types on one central place, to help avoid "forgetting" to initialize it

diesieben07 commented:
Updated as discussed on IRC.

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -156,90 +154,62 @@ public BiomeInfo(Type[] types)
+    public static void addTypes(Biome biome, Type... types)
     {
-        types = listSubTags(types);
+        Preconditions.checkArgument(ForgeRegistries.BIOMES.containsValue(biome), "Cannot register biome types for unregistered biome");
```
should be  "Cannot add biome types for unregistered biome"

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -21,133 +21,131 @@
+            return !subTags.isEmpty();
+        }
+
+        public String name()
```
should be `getName()`

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -156,90 +154,62 @@ public BiomeInfo(Type[] types)
 
     /**
-     * Checks to see if two biomes are registered as having the same type
+     * Checks to see if the two given biomes have tags in common.
```
should be types instead of tags here

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
-     * creation events.
+     * Automatically tags a given biome with appropriate types based on certain heuristics.
+     * If a biome's types are requested and the biome has not yet been tagged with any types, the biome's tags will be
+     * determined using this method
```
should be "determined and added using this method", since it actually adds the types too

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
+     * Ensure that the given biome has been tagged with at least one type
+     * @param biome the biome
+     */
+    static void ensureTagged(Biome biome)
```
package-private?
`ensureHasTypes` might be better, I want to avoid using both types and tags everywhere.

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
-        if(!isBiomeRegistered(biome))
+        if (!hasTypes(biome))
         {
+            FMLLog.warning("Biome %s has not been tagged with any types, types will be assigned on a best-effort guess.");
```
To avoid "tag", this could be "Biome %s does not have any types added, types will be assigned on a best-effort guess."

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
     }
 
-    private static Type[] listSubTags(Type... types)
+    private static List<Type> listSubTags(Type... types)
```
`listSubTypes`

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -21,133 +21,131 @@
+            byName.put(name, this);
         }
 
         private boolean hasSubTags()
```
`hasSubTypes`

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -21,133 +21,131 @@
+        private static final Map<String, Type> byName = new HashMap<String, Type>();
+
+        private final String name;
+        private final List<Type> subTags;
```
`subTypes`

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -156,90 +154,62 @@ public BiomeInfo(Type[] types)
 
     /**
-     * Registers a biome with a specific biome type
+     * Tags a biome with the given types.
```
To avoid "tag", use "Sets a biome with the given types." or "Adds the given types to a biome"

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -156,90 +154,62 @@ public BiomeInfo(Type[] types)
-     * @param types the types to register the biome as
-     * @return returns true if the biome was registered successfully
+     * @param biome the biome to be tagged
+     * @param types the types to tag with
```
these two parameter docs can be removed, they don't explain anything

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -156,90 +154,62 @@ public BiomeInfo(Type[] types)
-                {
-                    typeInfoList[type.ordinal()] = new ArrayList<Biome>();
-                }
+        List<Type> subTags = listSubTags(types);
```
`subTypes`

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -156,90 +154,62 @@ public BiomeInfo(Type[] types)
 
     /**
-     * Returns a list of biomes registered with a specific type
+     * Returns the set of biomes that have been tagged with the given type.
```
that have been set with

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -156,90 +154,62 @@ public BiomeInfo(Type[] types)
 
     /**
-     * Gets a list of Types that a specific biome is registered with
+     * Get the set of types that the given biome has been tagged with.
```
has been set with.

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -156,90 +154,62 @@ public BiomeInfo(Type[] types)
-     * @param biomeA
-     * @param biomeB
+     * @param biomeA the first biome
+     * @param biomeB the second biome
```
you can remove these two parameter docs

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
 
     /**
-     * Checks to see if the given biome is registered as being a specific type
+     * Checks to see if the given biome is has been tagged with the given type.
```
has been set with the given

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
-     * @return returns true if the biome is registered as being of type type, false otherwise
+     * @param biome the biome
+     * @param type  the type to check for
+     * @return true if the biome has been tagged with the type, false otherwise
```
has been set with the

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
 
     /**
-     * Checks to see if the given biome has been registered as being of any type
+     * Checks to see if the given biome has been tagged with any type.
```
has been set with any type

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
+     *
      * @param biome the biome to consider
-     * @return returns true if the biome has been registered, false otherwise
+     * @return true if the biome has been tagged with any type
```
@return true if the biome has has any types added

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
-     * This is called by Forge at postinit time. It will additionally dispatch any deferred decorator
-     * creation events.
+     * Automatically tags a given biome with appropriate types based on certain heuristics.
+     * If a biome's types are requested and the biome has not yet been tagged with any types, the biome's tags will be
```
If a biome's types are requested and the biome has not had any types added, the biome's types will be

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
-     * And another method requests information about it
-     *
-     * @param biome the biome to be considered
+     * @param biome the biome to be tagged
```
the biome to analyze and add types to

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
 
-    private static void checkRegistration(Biome biome)
+    /**
+     * Ensure that the given biome has been tagged with at least one type
```
Ensure that the given biome has had at least one type added

mezz replied:
missed this one

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
-    private static void checkRegistration(Biome biome)
+    /**
+     * Ensure that the given biome has been tagged with at least one type
+     * @param biome the biome
```
remove this parameter doc

mezz reviewed (changes requested):
alright I went crazy with the review but I don't like the way things read with "tag" and "type" together all over. I think it should be type only.

diesieben07 committed:
- change "tag" to "type"
- remove redundant javadocs

diesieben07 commented:
Updated.

mezz reviewed (changes requested):
Looks great, but you just missed one doc.
I asked @Adubbz for feedback when they have a chance to take a look.

diesieben07 committed:
fix one last javadoc

diesieben07 commented:
Fixed another Javadoc.

Adubbz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +214,231 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
      */
-    public static void registerAllBiomesAndGenerateEvents()
+    public static boolean hasTypes(Biome biome)
     {
```
I feel like this name is rather confusing given that hasType accepts a Type argument, thus implying that hasTypes should therefore be used to check whether a biome is assigned multiple types as provided by the modder.

Adubbz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +214,231 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
      */
-    public static void registerAllBiomesAndGenerateEvents()
+    public static boolean hasTypes(Biome biome)
     {
```
I feel like this name is rather confusing given that hasType accepts a Type argument, thus implying that hasTypes should therefore be used to check whether a biome is assigned multiple types as provided by the modder.

diesieben07 replied:
I was debating this as well, `hasAnyType` maybe?

Adubbz replied:
Yeah, I think that's better

diesieben07 committed:
hasTypes -> hasAnyType

mezz commented:
```
java.lang.ExceptionInInitializerError
	at net.minecraftforge.common.BiomeDictionary.registerVanillaBiomes(BiomeDictionary.java:403)
	at net.minecraftforge.common.BiomeDictionary.<clinit>(BiomeDictionary.java:156)
	at net.minecraftforge.common.ForgeModContainer.registerAllBiomesAndGenerateEvents(ForgeModContainer.java:445)
	at net.minecraftforge.common.ForgeModContainer.postInit(ForgeModContainer.java:431)
        ... snip
Caused by: java.lang.NullPointerException
	at net.minecraftforge.common.BiomeDictionary$Type.<init>(BiomeDictionary.java:99)
	at net.minecraftforge.common.BiomeDictionary$Type.<clinit>(BiomeDictionary.java:42)
	... 42 more
```

Tested and it crashes. 
`public static final Type HOT` 
is being statically created before 
`private static final Map<String, Type> byName`.

diesieben07 commented:
I'm an idiot.

diesieben07 committed:
fix static initializer ordering, add Type::toString

diesieben07 commented:
Fixed.

mezz commented on src/main/java/net/minecraftforge/common/BiomeDictionary.java:
```diff
@@ -249,251 +220,231 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
-        if(!isBiomeRegistered(biome))
+        if (!hasAnyType(biome))
         {
+            FMLLog.warning("No types have been added to Biome %s, types will be assigned on a best-effort guess.");
```
Alright I'm done trying to test this until you show me a screenshot in-game haha.
This line crashes.

bs2609 replied:
`biome` needs adding as a parameter, I assume

bs2609 commented on src/main/java/net/minecraftforge/common/ForgeModContainer.java:
```diff
@@ -428,10 +428,24 @@ public void preInit(FMLPreInitializationEvent evt)
+                decorator.fireCreateEventAndReplace(biome);
+            }
+
+            BiomeDictionary.ensureHasTypes(biome);
```
If we have to call through to `BiomeDictionary` anyway, what's the benefit of moving this method here over just making it package-private?

diesieben07 replied:
Good question.

diesieben07 committed:
Fix string formatting

diesieben07 commented:
I promise it works now.

mezz commented:
Tested it out and looks good to me.

LexManos commented:
This is nothing more then a code cleanup, I don't see any 'improvements' here. However we're still {getting close to the end} of our 1.11 update phase so I guess we can accept it.

diesieben07 commented:
It started out much different than it is now and i never bothered to change the title.

AEnterprise commented:
not sure if this is the right place to ask: but is there a list somewhere of all the breaking changes made in the 1.11 update in forge or should i find the PR's with 'breaking change' labels? because i'm not sure those include the breaking changes and pointers to new stuff made by the core forge team

mezz commented:
I'm trying to keep PRs labeled well, so [this a good start](https://github.com/MinecraftForge/MinecraftForge/issues?utf8=%E2%9C%93&q=label%3ABreakingChange%20is%3Amerged). 
There is no curated list of breaking changes, if you wanted to make one that might be nice.

AEnterprise commented:
i might try that later, getting back into modding but not always finding enough time due to school

but thanks for the answer, i'll make sure i ask future questions on the forums (sorry, only just realised that would be the right place to ask instead of here)

KingLemming commented:
So just to be clear, this actually is broken as hell.

getBiome(Type) always returns an empty list because in addTypes, the base biome is never actually added to the list of types. listSuperTypes always seems to return an empty list.

I feel like some logic got dropped or screwed up here - but the fact of the matter is that the mappings are hosed and there's no way to get a list of all biomes which have a specific type associated with them.

mezz commented:
File a bug or PR please, this is closed so comments will get lost easily.

KingLemming commented:
Planned on it. I just had to voice my displeasure. Haven't had time for a PR yet.

mezz commented:
Your somewhat dramatic public complaint has been noted and filed appropriately thanks :P

KingLemming commented:
<3
